### ISSUE TITLE
[Bug] Uninitialized rows reach the block-FP8 GEMMs with --moe-a2a-backend deepep (masked by DeepGEMM, NaN with flashinfer)

### ISSUE THREAD

**Issue by zhendonghua** [sgl-project/sglang#33106]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] Please use English.

### Describe the bug

Some rows entering the block-FP8 linears hold **uninitialized memory**. Every GEMM backend
computes on them and they only differ in what comes out. `deep_gemm` — the auto default on
Hopper — saturates the garbage to a finite value, so nothing has ever noticed. `flashinfer`
overflows it to NaN, `SGLANG_ENABLE_ASYNC_ASSERT` scans all of `next_token_logits`
including those rows, and the scheduler dies during startup:

```
Assertion `NaN detected! sampler: next_token_logits` failed.
```

So this is an SGLang initialization bug that the default backend has been masking, not a
flashinfer bug. Zeroing the rows makes the crash go away (A/B below), so the cause is
established — what I could not find is the allocation that produces them.

`SGLANG_ENABLE_ASYNC_ASSERT` is not a debug-only knob: `_pr-test-stage.yml` sets it to
`true` for every CI stage except `base-a`.

### Reproduction

`main` @ `89f4a80c1`, unmodified, 2x H200, `lmsys/sglang-ci-dsv3-test`. No DP attention, no
speculative decoding. **Fails 2/2:**

```bash
SGLANG_ENABLE_ASYNC_ASSERT=true python3 -m sglang.launch_server \
  --model-path lmsys/sglang-ci-dsv3-test --trust-remote-code \
  --tp 2 --moe-a2a-backend deepep \
  --fp8-gemm-backend flashinfer_deepgemm
```

The server never becomes healthy; it dies serving the 7-token startup request
(`Prefill batch, #new-seq: 1, #new-token: 7`). Changing only the last line to
`--fp8-gemm-backend deep_gemm` starts fine.

Which knobs matter, all with the assertion on:

| MoE a2a | DP attn | fp8 backend | result |
| --- | --- | --- | --- |
| `deepep` | off | `flashinfer_deepgemm` | **FAIL** |
| `deepep` | **on** | `flashinfer_deepgemm` | **FAIL** — DP attention is irrelevant |
| `none` | off | `flashinfer_deepgemm` | ok — `deepep` is the trigger |
| `deepep` | off | `deep_gemm` | ok |
| `deepep` | off | `triton` | ok |

### Root cause

**The rows are uninitialized, not miscomputed.** Printing the per-row `amax` of the input at
`forward_mha.py::forward_normal_core -> o_proj` (`M=8`, 7 real tokens), rows 0–6 stay in
`[0.09, 1.1]` while the last row reads, across calls at the same call site:

```
6.70e+16    5.09e+35    1.69e+38    2.30e+38    3.10e+38
```

A value that jumps 22 orders of magnitude call to call is stale memory, not arithmetic. The
attention kernel writes the 7 real rows; row 7 is never written. The same shows up at
`forward_normal_prepare` (`q_proj`, `kv_a_proj_with_mqa`, `kv_b_proj`) and at
`_forward_shared_experts`.

**Neutralizing those rows fixes the crash.** Same binary, same failing config, one env var.
The clamp sits inside the backend so it covers every call site (CUDA-graph safe, no host
sync):

```python
# immediately before the fp8_blockscale_gemm_sm90 call
if os.environ.get("FI_CLAMP") == "1":
    input_2d = torch.where(
        input_2d.abs().amax(dim=-1, keepdim=True) > 1e35,
        input_2d.new_zeros(()),
        input_2d,
    )
```

| `FI_CLAMP` | runs | server start |
| --- | --- | --- |
| `0` | 1/1 | **FAIL** |
| `1` | 2/2 | **ok** |

**Why only flashinfer trips it.** Injecting one row of a known magnitude into that shape
(N=2560, K=1024, block `[128, 128]`, real weights, 7 normal rows + 1 injected):

| injected row magnitude | `deep_gemm` | `triton` | `flashinfer_deepgemm` |
| --- | --- | --- | --- |
| <= 1e35 | clean | clean | clean |
| 1e36 – 3.2e37 | clean | **NaN + Inf** | clean |
| >= 1e38 | clean | clean | **NaN** |

bf16 max is 3.39e38, so all of these are representable inputs. `deep_gemm` is clean at every
magnitude tested, which is why the bug has been invisible. `triton` survives the server
repro only because the magnitudes this model happens to produce skip its window.

Two consequences beyond this one config:

- Any change to the auto-dispatch priority, or to how a backend saturates, turns a latent
  condition into a hard startup failure on a stock config.
- The assertion cannot distinguish a real NaN from one propagated out of a garbage row, so
  it reports a correctness failure where there is none. With the assertion off, `output_ids`
  from the flashinfer run are byte-identical to the `deep_gemm` run — generated tokens are
  correct today, but the stale data is carried through every layer.

### What I could not find

**Which allocation produces the rows.** The concrete gap: a 7-token prefill reaches the GEMM
as `M=8`, and nothing writes row 7. I did not find what pads 7 to 8, nor where that buffer
is allocated.

Best lead: with both CUDA-graph backends disabled the same config **passes**, and the garbage
magnitudes in that run topped out at `5.09e35` — below the flashinfer threshold. That run
also carried an instrumentation probe, so I cannot separate "graphs off" from "probe
present", but the CUDA-graph static buffers look like the place to start.

Ruled out along the way:

- **DP attention** — fails with it off, so these are not DP-attention padding rows.
- **The two `dp_attention.py` DP buffers** — `torch.zeros` at allocation, plus an explicit
  `buffer.zero_()` outside the `use_symmetric_memory` context, changes nothing.
- **A single-call-site patch** — zeroing only before `o_proj` leaves the crash in place;
  garbage enters through several linears. Worth knowing before retrying that route.
- **All-zero rows as the trigger** — a fully zeroed row is clean on all three backends, so
  this is not a `scale = amax / 448` division by zero. Zero is a safe value to write.
- **flashinfer leaving output elements unwritten** — would have explained NaN from a clean
  input. Poisoning the block the output tensor reuses, then running the GEMM: 0 poison
  elements survive out of 20480, across M=8/16/31 and K=1024/2048, on all three backends.

### Suggested fix

1. **Initialize the region at the source.** The A/B shows that neutralizing these rows is
   sufficient, so initializing them where they are allocated should be too. The in-backend
   clamp above is a diagnostic, not a proposal — it pays a `where` on every GEMM and treats
   the symptom.
2. **Restrict the NaN assertion to valid rows.** Independently correct: the sampler knows
   which rows belong to real requests, and scanning the rest produces exactly this kind of
   false report.

Happy to write either once someone can point me at the buffer.

### Notes on attribution

The repro, the knob table's first row, the `FI_CLAMP` A/B, the magnitude table and the
poison test are all on `main` @ `89f4a80c1`. The remaining knob-table rows and the per-row
`amax` trace come from a tree ~430 commits behind that main, where every file on the failing
path differs — read those as isolating the *config* variables only.

One anomaly I cannot explain: a single instrumented call showed NaN in the flashinfer output
while all eight input rows read finite and bounded (`0.097`–`0.473`). The poison test rules
out the mechanism I found most plausible. The `FI_CLAMP` A/B outweighs it as evidence about
the cause, so I record it as an anomaly rather than a competing hypothesis.

### Environment

```
Python: 3.12.3 | PyTorch: 2.11.0+cu130 | CUDA 13.0 | Driver 580.105.08
GPU: 8x NVIDIA H200 (SM 9.0), NV18 all-to-all, single node
sglang: 0.5.14 (main @ 89f4a80c1) | sglang-kernel: 0.4.5
flashinfer_python / _cubin: 0.6.12 | flashinfer_jit_cache: 0.6.12+cu130
triton: 3.6.0 | transformers: 5.8.1
```

**Comment by zhendonghua** [comment]
Updated the description: the causal claim is now backed by a controlled A/B rather than inference.

Zeroing the out-of-range rows inside `flashinfer_deepgemm_w8a8_block_fp8_linear_with_fallback` (so it covers every call site, no host sync, CUDA-graph safe) makes the same binary on the same failing config start cleanly 2/2, versus FAIL with the clamp off. Details in the new "Confirming the cause" section.

Also added: an earlier single-call-site patch that does *not* work (garbage enters through several linears, not just `o_proj`), and a poison test refuting the alternative explanation that flashinfer leaves part of its output tensor unwritten — all three backends write every output element.
